### PR TITLE
Notice: Add unit tests, Round 2

### LINKED ISSUE #7941: Components: Add tests for <Notice /> component
The [`<Notice />` component](https://github.com/Automattic/wp-calypso/tree/master/client/components/notice) is currently not tested. We should seek to add tests to verify expected render behavior of the component. [Enzyme](http://airbnb.io/enzyme/) (and [`chai-enzyme` assertions](https://github.com/producthunt/chai-enzyme)) may be useful utilities for rendering and asserting expected render results. Ideas for test cases may include some or all of the following:

- Render returns expected element
 - Optionally rendered components are or are not present
 - Modifier classes by incoming props or component state
- User interactions have expected results
- Impact of lifecycle methods when props or state change during lifetime of component

When submitting a pull request, reference this issue.

### PR DESCRIPTION
**Pull request by jasmussen**
This is a fresh pull request containing a merged, rebased version of the code in https://github.com/Automattic/wp-calypso/pull/8536. That old PR was reverted due to errors in devdocs on production, but the original intention was to fix #7941.

The error occurred on the [UI Components docs page](https://wpcalypso.wordpress.com/devdocs/design), and looked like this: 

```
Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object. Check the render method of 'Notices'
```

This PR contains the same issue, and some more warnings:

```
warning.js:36 Warning: React.createElement: type should not be null, undefined, boolean, or number. It should be a string (for DOM elements) or a ReactClass (for composite components). Check the render method of `Notices`.
```

Error can be tested on http://calypso.localhost:3000/devdocs

Hoping to fix the error and reinstate the Notice component translate refactor, as well as the tests. CC: @aduth and because you expressed interest yesterday, @lamosty.

### CODE REVIEW

**Comment by lamosty** [comment]
Alright, I found the issue. :)

In [notice/docs/example](https://github.com/Automattic/wp-calypso/blob/master/client/components/notice/docs/example.jsx#L11-L11), we import the notice like this:

```js
Notice = require( 'components/notice' );
```

However, in [notice/index](https://github.com/Automattic/wp-calypso/blob/624d4486c8f4eb1e2e269f2907044f84614041b3/client/components/notice/index.jsx#L14-L14), we export both a named variable `Notice` and a **default** export.

Using ES6 to import these would look like:

```js
import Notice from 'components/notice';
```

The above would automatically import the **default** export from the `notice/index` file. If we wanted to import both the default export and the named variable in ES6, we could do it like:

```js
import LocalizedNotice, { Notice } from 'components/notice';
```
`LocalizedNotice` would contain the **default** export and `Notice` would contain the `export cost Notice = ...`.

So, getting to point. The main problem is that ES5 `require` method, which is used in [notice/docs/example](https://github.com/Automattic/wp-calypso/blob/master/client/components/notice/docs/example.jsx#L11-L11) as described above, behaves a bit differently. Doing this:

```js
var Notice = require( 'components/notice' );
```

results in `Notice` being an object containing these props: `default: [ the localized Notice ], Notice: [ the named export Notice ]`.

As we can see, that's why React complains that `var Notice` is not a string or a React component. It's just an object with exports.

So, the fix is to either use ES6 imports as described above (recommended) or fix it by importing the localized `Notice` component like this:

```js
var Notice = require( 'components/notice' ).default;
```

**Comment by jasmussen** [comment]
Wow, amazing investigative work @lamosty, thanks a bunch. Before I do any code changes, I'll let this sit to gather opinions from the stronger JavaScripters than myself (Andrew?) — as I have no strong opinion on what would be the better long term approach.

**Comment by lamosty** [comment]
No problem. :)

> what would be the better long term approach

Definitely going the ES6 route (ie using `import Notice from 'components/notice'`).

If you prefer to wait for Andrew, alright. :)

**Comment by ockham** [comment]
> or fix it by importing the localized `Notice` component like this:

> ```js
var Notice = require( 'components/notice' ).default;
```

On a sidenote, for a module **only** containing a default export (and no other exports), this wouldn't even be necessary, since we have a [Babel plugin](https://github.com/Automattic/wp-calypso/blob/a934a06c65889890f82f26ebd1202aabb57cd72d/package.json#L19) to automate that for us.

**Comment by aduth** [comment]
The docs for that Babel plugin provide some good context too:

https://github.com/59naga/babel-plugin-add-module-exports

Sorry I hadn't caught this myself in my review. Not something I would have expected to cause breakage, though makes total sense after thinking on it.

I agree ES6 is the way to go, though I anticipate there's probably quite a few places we're importing Notice via `require`.

**Comment by jasmussen** [comment]
Alright, seems like the ES6 route is the accepted way.

So just to be clear, that means grepping the entire codebase looking for `Notice = require( 'components/notice' );` and replacing with `import Notice from 'components/notice')`, then testing thoroughly that nothing's broken — does that sound right?

**Comment by aduth** [comment]
Minus the typo on the parenthesis, but yeah 😄

**Comment by jasmussen** [comment]
> Minus the typo on the parenthesis

😂

**Comment by jasmussen** [comment]
Rebased, and replaced any instances of `Notice = require( 'components/notice' );` that I could find with `import Notice from 'components/notice';`. In testing Calypso afterwards, I couldn't reproduce the error messages that the previous merge generated.

Would very much appreciate a good look at the code: https://github.com/Automattic/wp-calypso/pull/8667/commits/31d81dbf9214e95a0d2ffcc48be0e833a561c7e2

**Comment by aduth** [comment]
I still see many more CommonJS `require` for notice:

- `client/components/domains/map-domain-step/index.jsx`
- `client/me/reauth-required/index.jsx`
- `client/me/security-2fa-backup-codes-prompt/index.jsx`
- `client/me/security-2fa-code-prompt/index.jsx`
- `client/me/security-2fa-enable/index.jsx`
- `client/me/security-2fa-setup-backup-codes/index.jsx`
- `client/me/security-2fa-sms-settings/index.jsx`
- `client/me/security-checkup/manage-contact.jsx`
- `client/my-sites/draft/index.jsx`
- `client/my-sites/upgrades/domain-management/add-google-apps/add-email-addresses-card.jsx`
- `client/my-sites/upgrades/domain-management/site-redirect/notice.jsx`
- `client/my-sites/upgrades/map-domain/index.jsx`
- `client/post-editor/editor-location/index.jsx`
- `client/reader/post-errors/index.jsx`
- `client/signup/steps/domains/index.jsx`

**Comment by aduth** [comment]
In case it's helpful, the regex search I did:

```bash
grep -rIE "'components/notice'[^;]" .
```

**Comment by jasmussen** [comment]
That was very helpful, thanks! I used Atoms multi file search.

**Comment by jasmussen** [comment]
Updated the files found using `grep -rIE "'components/notice'[^;]" .` (except for a post-editor test which it also found), and rebased. Tests well with no errors.

**Comment by klimeryk** [comment]
I'm wondering if this way of fix this issue is not too fragile/magical - I've just stumbled upon this PR after having a conflict in a file I was working on. Adding one ES6 style import and leaving the rest of them as CommonJS seems like a very temporary and fragile solution. I can imagine that someone else comes along and will revert that one `import` to use `require` to be consistent with the rest of the file. Not to mention that it's very unclear _why_ that one `import` has been added and what's the significance of it. I had always assumed that whether I'm using `require` or `import` it's all the same in the end - I'm sure I'm not the only one :)

TL;DR - a less fragile solution would be very much appreciated :) Especially since nothing is explicitly preventing other developers from `require`'ing the `Notice` component again (for example, in existing code that is depending on `require` to pull in modules).

**Comment by jasmussen** [comment]
Should I revert until we think of a less "fragile" implementation?

**Comment by aduth** [comment]
>I can imagine that someone else comes along and will revert that one `import` to use `require` to be consistent with the rest of the file.

Nobody should be doing this. Nobody should be introducing any new `require` calls for that matter (outside of Webpack configs and server code).

It's worth highlighting that the way it was previously took advantage of our own hack introduced to ease the upgrade to Babel 6 ([`babel-plugin-add-module-exports`](https://www.npmjs.com/package/babel-plugin-add-module-exports)).

>Adding one ES6 style import and leaving the rest of them as CommonJS seems like a very temporary and fragile solution.

Temporary, yes. Eventually it should all be `import`, but migrating all of a file's imports in same the pull request can be error-prone and requires significantly more review effort.

To my earlier point, even if it's inconsistent with the rest of the imports, if I'm introducing a new dependency, I'll always use `import`. Otherwise the file will likely never be migrated. A more incremental approach helps move us in that direction gradually.

>I had always assumed that whether I'm using require or import it's all the same in the end - I'm sure I'm not the only one :)

It's true that there are subtle and breaking differences, particularly around when a file has both named and default exports (as was the case here). Another is an implementation detail of Babel, which is that imports are always moved to the top of the file, which has come up at least once in #7038. Part of this is education around the import semantics, or in the second case, coming up with other solutions to address fragile load order dependencies (e.g. #7232).

**Comment by klimeryk** [comment]
Thank you for your input, @aduth. I'm sure you know this ecosystem and project way better than me :) But there are many developers working on this, with different levels of knowledge of JS and Calypso and you can't expect everyone to know the difference between `require` and `import`. You say that nobody should be adding new `require` calls and that mixing `import` and `require` is OK. But nothing is preventing the first (no ESLint rule, no mention in our coding guidelines, etc.), while the latter I'd say is discouraged by the rule `Try to stay consistent with the rest of the codebase`.
But again, I don't want to draw out this discussion (as it often does in these kind of cases). I wonder if it would be possible to add a new custom rule to our ESLint ones that would throw an error when someone tries to `require` `components/notice`? That would address my issue with this PR completely :) (I'd have suggested that in my first comment, but the idea just came to me :))

**Comment by aduth** [comment]
>But again, I don't want to draw out this discussion (as it often does in these kind of cases).

These discussions are very valuable, so please don't hesitate to be thorough in expressing your concerns!

I agree it can be confusing, and that there should be a goal of general consistency. But from my own experience, unless there's an incremental approach that can be included in arguably unrelated pull requests (like this one), very little effort is put toward adopting new best practices. And in the end, this itself leads to confusion for new developers because those developers often reference the files they see, and if those files include old syntax, they'd be misled in thinking that those are the best practices.

I also agree that tooling, particularly ESLint, can help us in drawing attention to code which doesn't adhere to our current best practices. However, there's a balance here in signal to noise ratio, since if we overwhelm a developer with warnings, there's a worry they may become accustomed to ignoring all of them instead of paying close attention. In this case, though, I think a `require` warning would be useful, though we'd want to limit it to the `client` directory.
